File: app/src/main/java/com/bogdanorzea/popularmovies/fragment/SettingsFragment.java
package com.bogdanorzea.popularmovies.fragment;

import android.content.ContentValues;
import android.content.DialogInterface;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.support.v7.app.AlertDialog;
import android.support.v7.preference.ListPreference;
import android.support.v7.preference.Preference;
import android.support.v7.preference.PreferenceFragmentCompat;
import android.support.v7.preference.PreferenceGroup;
import android.support.v7.preference.PreferenceScreen;

import com.bogdanorzea.popularmovies.R;
import com.bogdanorzea.popularmovies.data.MoviesContract;

public class SettingsFragment extends PreferenceFragmentCompat
        implements SharedPreferences.OnSharedPreferenceChangeListener {

    @Override
    public void onCreatePreferences(Bundle savedInstanceState, String rootKey) {
        addPreferencesFromResource(R.xml.preferences);

        PreferenceScreen preferenceScreen = getPreferenceScreen();
        int preferenceCount = preferenceScreen.getPreferenceCount();

        for (int i = 0; i < preferenceCount; i++) {
            Preference preference = preferenceScreen.getPreference(i);
            setSummary(preference);
        }


        Preference button = findPreference(getString(R.string.pref_clear_favorites_key));
        button.setOnPreferenceClickListener(new Preference.OnPreferenceClickListener() {
            @Override
            public boolean onPreferenceClick(Preference preference) {
                DialogInterface.OnClickListener dialogClickListener = new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        if (which == DialogInterface.BUTTON_POSITIVE) {
                            ContentValues values = new ContentValues();
                            values.put(MoviesContract.MovieEntry.COLUMN_NAME_FAVORITE, 0);

                            getContext().getContentResolver().update(MoviesContract.CONTENT_URI, values, null, null);
                        }

                        dialog.dismiss();
                    }
                };


                AlertDialog.Builder builder = new AlertDialog.Builder(getContext());
                builder.setMessage(R.string.confirm_clear_favorites)
                        .setPositiveButton(R.string.button_yes, dialogClickListener)
                        .setNegativeButton(R.string.button_no, dialogClickListener)
                        .show();

                return true;
            }
        });
    }

    private void setSummary(Preference preference) {
        if (preference instanceof PreferenceGroup) {
            PreferenceGroup preferenceGroup = (PreferenceGroup) preference;
            for (int i = 0; i < preferenceGroup.getPreferenceCount(); i++) {
                setSummary(preferenceGroup.getPreference(i));
            }
        }

        if (preference instanceof ListPreference) {
            ListPreference listPreference = (ListPreference) preference;
            preference.setSummary(listPreference.getEntry());
        }
    }

    @Override
    public void onStart() {
        super.onStart();
        getPreferenceScreen().getSharedPreferences().registerOnSharedPreferenceChangeListener(this);
    }

    @Override
    public void onStop() {
        super.onStop();
        getPreferenceScreen().getSharedPreferences().unregisterOnSharedPreferenceChangeListener(this);
    }

    @Override
    public void onSharedPreferenceChanged(SharedPreferences sharedPreferences, String preferenceKey) {
        Preference preference = findPreference(preferenceKey);
        if (null != preference) {
            setSummary(preference);
        }
    }
}

File: app/src/main/java/com/bogdanorzea/popularmovies/ui/settings/SettingsFragment.kt
package com.bogdanorzea.popularmovies.ui.settings

import android.content.ContentValues
import android.content.DialogInterface
import android.content.SharedPreferences
import android.os.Bundle
import android.support.v7.app.AlertDialog
import android.support.v7.preference.ListPreference
import android.support.v7.preference.Preference
import android.support.v7.preference.PreferenceFragmentCompat
import android.support.v7.preference.PreferenceGroup

import com.bogdanorzea.popularmovies.R
import com.bogdanorzea.popularmovies.data.MoviesContract

class SettingsFragment : PreferenceFragmentCompat(), SharedPreferences.OnSharedPreferenceChangeListener {

    override fun onCreatePreferences(savedInstanceState: Bundle?, rootKey: String?) {
        addPreferencesFromResource(R.xml.preferences)

        for (i in 0 until preferenceScreen.preferenceCount) {
            setSummary(preferenceScreen.getPreference(i))
        }

        findPreference(getString(R.string.pref_clear_favorites_key))?.setOnPreferenceClickListener {
            val dialogClickListener = { dialog: DialogInterface, which: Int ->
                when (which) {
                    DialogInterface.BUTTON_POSITIVE -> {
                        val values = ContentValues()
                        values.put(MoviesContract.MovieEntry.COLUMN_NAME_FAVORITE, 0)

                        context?.contentResolver?.update(MoviesContract.CONTENT_URI, values, null, null)
                    }
                }

                dialog.dismiss()
            }

            context?.let {
                AlertDialog.Builder(it)
                        .setMessage(R.string.confirm_clear_favorites)
                        .setPositiveButton(R.string.button_yes, dialogClickListener)
                        .setNegativeButton(R.string.button_no, dialogClickListener)
                        .show()
            }

            true
        }
    }

    private fun setSummary(preference: Preference) {
        when (preference) {
            is PreferenceGroup -> {
                for (i in 0 until preference.preferenceCount)
                    setSummary(preference.getPreference(i))
            }
            is ListPreference -> preference.setSummary(preference.entry)
        }
    }

    override fun onStart() {
        super.onStart()
        preferenceScreen.sharedPreferences.registerOnSharedPreferenceChangeListener(this)
    }

    override fun onStop() {
        super.onStop()
        preferenceScreen.sharedPreferences.unregisterOnSharedPreferenceChangeListener(this)
    }

    override fun onSharedPreferenceChanged(sharedPreferences: SharedPreferences?, preferenceKey: String?) {
        findPreference(preferenceKey)?.let {
            setSummary(it)
        }
    }
}
